channeldb/payment_control.go

```diff
@@ -359,15 +517,29 @@ func nextPaymentSequence(tx *bbolt.Tx) ([]byte, error) {
 // fetchPaymentStatus fetches the payment status of the payment. If the payment
 // isn't found, it will default to "StatusUnknown".
 func fetchPaymentStatus(bucket *bbolt.Bucket) PaymentStatus {
-	if bucket.Get(paymentSettleInfoKey) != nil {
-		return StatusSucceeded
+	htlcsBucket := bucket.Bucket(mppHtlcsBucket)
+	if htlcsBucket != nil {
+		htlcs, err := fetchHTLCAttempts(htlcsBucket)
+		if err != nil {
+			return StatusUnknown
+		}
+
+		// Go through all HTLCs, and return StatusSucceeded if any of
+		// them did succeed.
+		// TODO(halseth): Is this iteration a perf bottleneck?
```

Comment:
I wouldn't be worried about that at this stage. Once this is the bottleneck, adoption has happened :)

Reply:
It is `O(N^2)` where N is the number of HTLCs sent for a payment I believe. So you don't need that may failed attempt before this becomes a problem.

Reply:
You mean that we are redetermining the payment status for every htlc result that comes in? I still think that this isn't a problem in practice. Do you have a number in mind when this really becomes problematic? Probably the fetch call is a lot hotter than this in memory loop.